Add a feature to allow an editor to embargo a project's files

briangow opened the pull request:
This PR adds an option to embargo / hide a project's files during submission via the Embargo tab. The editor can choose the number of days (up to 365) to embargo a project's files for (from the publication date). The project can then be published but the files will not be accessible to users during the embargo. A future PR will add the ability to make the files available only to certain users while they are under embargo. 

An option for removing the embargo is also provided for a published project via the Manage Project page.

I have been testing this using this workflow on my local database:
1. Login as admin
2. Create a new version of "Demo for visualizing waveforms"
3. Admin console - assign admin as editor - edit project - Embargo tab (this can be done at the other submission stages also, not just during edit project) - enter number of days to embargo.
4. Publish the project
5. Admin console - projects - published - Manage the published version - Remove embargo (2/3 way down the page)

bemoody commented:
> pytz.UTC.localize(datetime.datetime.utcnow())

Use `timezone.now()` (`from django.utils import timezone`)

bemoody commented:
We also want to enforce download restrictions (you can't download the files even if you know where they are.)  Probably the best thing for now is for `sign_dua` (project/views.py) to reject requests for projects under embargo.

Eventually we'll want to add some escape hatch for authorized users/organizations.

briangow commented:
Thanks @bemoody ! I've updated the `.now` time call as suggested. I agree we should prevent file downloads in the cases where the user might be able to guess the path. I have added a `project.embargo_active()` constraint to `sign_dua`, as suggested. When the embargo is active this will prevent a user from signing the DUA and presumably from downloading the files even if they know the path to them.

tompollard commented:
> We also want to enforce download restrictions (you can't download the files even if you know where they are.) Probably the best thing for now is for sign_dua (project/views.py) to reject requests for projects under embargo.

Does this approach solve the issue for open access projects? I haven't looked at the code, but it sounds like it would only work for projects that require a DUA.

briangow commented:
@tompollard , I don't believe it would solve the problem for open-access projects. Do we ever anticipate embargoing files on open access projects? If so, we'd have to think of another solution. 

I did confirm that at least locally the approach above works if there is a DUA required for the project.

tompollard commented:
> Do we ever anticipate embargoing files on open access projects?

Yes, I think it is likely that open access projects would be embargoed (e.g. for the MIMIC waveforms).

bemoody commented:
Two sides to that question:

- Can we have a project that is restricted up until a certain date, and then becomes open?  It's possible of course, but there is not currently any way to do that.

- *Should we* display a green "Open Access" banner on projects that are not yet openly accessible?

rgmark commented:
When a dataset is embargoed it is not 'open access' right? It becomes 
open access then embargo ends?

Roger

On 12/12/2022 4:21 PM, Tom Pollard wrote:
>
>     Do we ever anticipate embargoing files on open access projects?
>
> Yes, I think it is likely that open access projects would be embargoed 
> (e.g. for the MIMIC waveforms).
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/MIT-LCP/physionet-build/pull/1740#issuecomment-1347346436>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABYOOSR6IVHIVEVRL5RZHZTWM6JMLANCNFSM6AAAAAASZQXOV4>.
> You are receiving this because you are subscribed to this 
> thread.Message ID: 
> ***@***.***>
>

-- 
Roger G. Mark, M.D., Ph.D.
Distinguished Professor in Health Sciences and Technology, Post-tenure
Institute of Medical Engineering and Science
Room E25-505
MIT
Cambridge, MA 02139
Tel 617-253-7818
Cell 617-359-2429
Fax 617-258-7859

tompollard commented:
> When a dataset is embargoed it is not 'open access' right? It becomes open access then embargo ends?

I agree, I think it's a true that a project isn't open access until the embargo ends. It may be desirable to indicate in some way that the project will become open access after the embargo.

> Can we have a project that is restricted up until a certain date, and then becomes open? It's possible of course, but there is not currently any way to do that.

I think this is the ultimate goal. A project that is "under embargo" may actually be accessible to a set of users (e.g. those members who belong to the consortium).

> Should we display a green "Open Access" banner on projects that are not yet openly accessible?

Hmm, I think on Roger's point above, probably not. It would be confusing to have an "open access" message and embargo displayed together. Probably best just to display an "Under embargo until XXXX" icon where the "open access" etc icons are normally displayed.

bemoody commented:
Conceptually I agree that it would be best to have the project automatically become open when the embargo ends.  There are a lot of details that need sorting out, beyond the purely technical.  For example, what license text is displayed, & what obligations do we have concerning third party licenses.

I'm inclined to start by adding this feature *only* for restricted/credentialed projects and extending to open access projects later.

bemoody commented:
So I think to avoid confusion, *for now*, the form should not allow the embargo date to be set if the access policy is open.

briangow commented:
> So I think to avoid confusion, _for now_, the form should not allow the embargo date to be set if the access policy is open.

This makes sense to me. If everyone agrees, I'll add a restriction that won't only allow the project to be embargoed if it is open access. 

I'll also explore options for how to restrict backend file access for open access projects.  Perhaps we can require `embargo_file_days=None` for file downloads, similar to how we currently require a signed DUA for restricted projects for file downloads. I can implement a solution for this in a future pull request.

briangow commented:
[18b0bae](https://github.com/MIT-LCP/physionet-build/pull/1740/commits/18b0bae605dae926ead96cf792e3fdb0b010e56e) removes the button to embargo an open access project's files and replaces it with this statement: "The files for an open access project cannot be placed under embargo".

briangow commented:
I implemented the above restriction to not allow the files for an open access project to be embargoed in order to prevent the issue described in https://github.com/MIT-LCP/physionet-build/issues/1744 . Typically you have to go through Django to gain access to a projects files but for open access projects there is a way to access them without going through Django. Therefore, marking an open access projects files as deprecated or under embargo will not necessarily prevent access. 

The fix for #1744 may be simpler than what is required for preventing access to files that are under embargo. We plan to introduce a feature which will allow certain groups (ex: consortium) to access a projects files that are under embargo. See https://github.com/MIT-LCP/physionet-build/issues/1745 . Therefore, we can't simply prevent read access to the static path on an open access project under embargo like we can with deprecated files. A more dynamic solution will be needed to prevent access to files in open access projects which are under embargo but can be accessed by the consortium, etc.

briangow commented:
For clarity, I anticipate at least three pull requests before the embargo files feature is fully released:

1. This PR, #1740, will allow editors to embargo a projects files for a set number of days on any projects which aren't open access. Once a project is published, this embargo can be ended sooner from the Manage Project page. 
2. Another PR will allow the editor to embargo files for projects which are open access. This will likely be related to the fix for #1744 but may be more involved as described above. 
3. Another PR will allow for certain groups/emails (ex: consortium) to access files in projects which are under embargo. 

@bemoody , @tompollard , I think 1. above is ready but please let me know if you think it requires any additional tweaks.

bemoody commented:
In console/templates/console/submission_info_card.html, the form fields
```
+            <form action="{% url 'submission_info' project.slug %}" method="POST" id="embargo_files_days_form">
+              {% csrf_token %}
+              {% include "form_snippet_no_labels.html" with form=embargo_files_days_form %}
```
would be better inside the modal box I think.

The form allows me to enter a negative number (or zero, which seems to do nothing.)  The submission info card also doesn't say how many days the project is embargoed, or allow me to undo it or change it if I make a mistake.  The project is *not published* at this point; changing embargo days while it is under submission isn't a problem.

(Also I haven't tested this, but suspect that if you submit + publish one project version under embargo, it will then be impossible to change the embargo period for a later version.)

I also would add a message to project_preview.html, e.g.
```
--- a/physionet-django/project/templates/project/project_preview.html
+++ b/physionet-django/project/templates/project/project_preview.html
@@ -221,6 +221,11 @@
     <!-- /.sidebar -->
   </div>
   <h2>Files</h2>
+  {% if project.embargo_files_days %}
+    <div class="alert alert-warning col-md-8" role="alert">
+      The files in this project version are under embargo until {{ project.embargo_files_days }} days after publication.
+    </div>
+  {% endif %}
   {% if is_lightwave_supported %}
     {% if project.has_wfdb %}
       <p><a href="{% url 'lightwave_project_home' project.slug %}?db={{ project.slug }}"><i class="fas fa-chart-line"></i> Visualize waveforms</a></p>
```

briangow commented:
@bemoody, thanks for the suggestions!

> In console/templates/console/submission_info_card.html, the form fields would be better inside the modal box I think.

I have moved the form to the modal. 

> The form allows me to enter a negative number (or zero, which seems to do nothing.) The submission info card also doesn't say how many days the project is embargoed, or allow me to undo it or change it if I make a mistake. The project is not published at this point; changing embargo days while it is under submission isn't a problem.

I have updated the form to only allow values between 0 and 365. Now 0 will reset `embargo_files_days` to `None`. When an embargo has been set, I now report what it is currently set to. I also allow the editor to change the number of days throughout the editing process. 

> (Also I haven't tested this, but suspect that if you submit + publish one project version under embargo, it will then be impossible to change the embargo period for a later version.)

You can now set the number of days under embargo independently across project versions. You can also independently choose to not embargo or break an embargo on a given version.

> I also would add a message to project_preview.html, e.g.

I have done this. I also added another sentence to make it clear that while we show the files in the preview they will be hidden once published.

tompollard commented on physionet-django/project/templates/project/project_preview.html:
```diff
@@ -221,6 +221,11 @@ <h5 class="card-header">Corresponding Author</h5>
   <h2>Files</h2>
+  {% if project.embargo_files_days %}
+    <div class="alert alert-warning col-md-8" role="alert">
+      The files in this project version are under embargo until {{ project.embargo_files_days }} days after publication. The files shown below will be hidden in the published project as part of this embargo.
```
nitpick but you can use the `|pluralize` tag here for day vs days. I think the syntax is:

```
day{{ project.embargo_files_days|pluralize }} after publication
```

briangow replied:
I see, thanks for the tip! I can implement this.

briangow replied:
This has been implemented in the templates.

tompollard commented on physionet-django/console/views.py:
```diff
@@ -274,6 +275,17 @@ def submission_info(request, project_slug):
+                project.embargo_files_days = None
+            else:
+                project.embargo_files_days = embargo_days_from_form
+            project.save()
```
Perhaps add a "Embargo set" message here? e.g. something like:

```
messages.success(request, f'An embargo was set for {project.embargo_files_days} day/s.')
```

briangow replied:
Ok, sure.

briangow replied:
I've added this message.

tompollard commented on physionet-django/console/views.py:
```diff
@@ -274,6 +275,17 @@ def submission_info(request, project_slug):
+            raise ServiceUnavailable()
+        elif embargo_files_days_form.is_valid():
+            embargo_days_from_form = embargo_files_days_form.cleaned_data['embargo_files_days']
+            if embargo_days_from_form == 0:
```
instead of `embargo_days_from_form`, just `days` might be more readable? it's clear from the context that we are talking about an embargo.

i'd say the same for `embargo_files_days_form` too i think. something like `embargo_form` seems nicer.

briangow replied:
I agree for the most part but we were trying to keep this as specific as possible to the files being under embargo. Therefore I keep using the word files in most places. As I think you mentioned, we may want to embargo additional content in the future. 

Perhaps: 
`embargo_days_from_form` -> `files_days`
`embargo_files_days_form` -> `embargo_files_form` or `files_form` (though this could easily be used elsewhere)

tompollard replied:
I know the post is `embargo_files`, so I don't need to be told in every subsequent variable that we are talking about embargo of the files. I just need to know that the variable contains `days` etc. 

Adding the extra information to the variable name interferes with readability for me, but I'm okay with whatever you end up with after giving it some thought. 

It's true that I think we should keep in mind that we may extend the embargo functionality to allow descriptions to be embargoed too, but (in most places) I don't think general language is counter to this. 

(e.g. an "embargo form" for me seems fine, because my guess is we would end up extending this form in the event we do allow embargoed descriptions).

briangow replied:
@tompollard I've changed to the variable names you suggested. If we add the ability to embargo the project description we can always revisit this if necessary.

tompollard commented on physionet-django/console/views.py:
```diff
@@ -259,6 +259,7 @@ def submission_info(request, project_slug):
 
     data = request.POST or None
     reassign_editor_form = forms.ReassignEditorForm(user, data=data)
+    embargo_files_days_form = forms.EmbargoFilesDaysForm()
```
see message below about variable names. `embargo_form` might be more readable.

tompollard reviewed (approved):
This looks good to me. I've added a couple of minor comments inline, and also added some notes below:

1. I like the use of the new "embargo" tab in the editorial console. If an embargo has been set, I think it should be much more obvious (otherwise someone could forget, or not realise if the project was transferred to them by another editor). I'd suggest putting an "Embargo" field under the "Project information" tab (which is the default view).
 
<img width="684" alt="Screen Shot 2022-12-16 at 3 01 57 PM" src="https://user-images.githubusercontent.com/822601/208183718-3f53d220-a32a-4c7f-9819-a6541d9d7c4b.png">

2. During copyediting, if I change a project that has an embargo from Restricted/Credentialed to Open, the form says "Invalid submission. See errors below." but no error is displayed. The "an open access project cannot be placed under embargo" error is actually hidden away on the "Embargo" tab. If possible, this error should be displayed to the editor somewhere more obvious, otherwise they will be left wondering why they cannot submit the form.

briangow commented:
> I like the use of the new "embargo" tab in the editorial console. If an embargo has been set, I think it should be much more obvious (otherwise someone could forget, or not realise if the project was transferred to them by another editor). I'd suggest putting an "Embargo" field under the "Project information" tab (which is the default view).

@tompollard , this is a good suggestion, thanks! I'll look into adding information about the embargo in the "Project information" tab.

> During copyediting, if I change a project that has an embargo from Restricted/Credentialed to Open, the form says "Invalid submission. See errors below." but no error is displayed. The "an open access project cannot be placed under embargo" error is actually hidden away on the "Embargo" tab. If possible, this error should be displayed to the editor somewhere more obvious, otherwise they will be left wondering why they cannot submit the form.

I had also noticed that you can't change a project from Restricted/Credentialed to Open. I haven't looked into why. I think this is independent of this PR though. I can look into it and open a new issue as needed.

briangow commented:
> I like the use of the new "embargo" tab in the editorial console. If an embargo has been set, I think it should be much more obvious (otherwise someone could forget, or not realise if the project was transferred to them by another editor). I'd suggest putting an "Embargo" field under the "Project information" tab (which is the default view).

> @tompollard , this is a good suggestion, thanks! I'll look into adding information about the embargo in the "Project information" tab.

I have added information about the Embargo status to the "Project information" tab.

> During copyediting, if I change a project that has an embargo from Restricted/Credentialed to Open, the form says "Invalid submission. See errors below." but no error is displayed. The "an open access project cannot be placed under embargo" error is actually hidden away on the "Embargo" tab. If possible, this error should be displayed to the editor somewhere more obvious, otherwise they will be left wondering why they cannot submit the form.

> I had also noticed that you can't change a project from Restricted/Credentialed to Open. I haven't looked into why. I think this is independent of this PR though. I can look into it and open a new issue as needed.

The message "The files for an open access project cannot be placed under embargo." under the embargo tab is not directly related to the error message "Invalid submission. See errors below.". However, once the bug preventing projects from being changed to Restricted / Open from Credentialed / Restricted ( #1758 ) is fixed we should have a process in place to handle attempts to switch a project under embargo to open. I've created an issue ( #1759 ) for this.

tompollard commented:
Thanks @briangow, looks good to me (with the existing issues that we're aware of, e.g. how to apply to "open access" projects, which we can deal with later).

A couple of other things that came to mind are:

- At the moment, the embargo doesn't seem to address data in the cloud (e.g. it looks like it may be possible to upload embargo data to cloud buckets from the "manage publish project" page in the console and provide access that way). 
- I think this is an unrelated issue, but there is a really annoying bug that after editing an embargo (and perhaps other fields) drops the editor into a "project info" page that doesn't include editorial forms.

briangow commented:
@tompollard , in regard to your points above:

> - At the moment, the embargo doesn't seem to address data in the cloud (e.g. it looks like it may be possible to upload embargo data to cloud buckets from the "manage publish project" page in the console and provide access that way).

I opened an issue to make sure this is considered if we allow open-access project files to be embargoed in the future: https://github.com/MIT-LCP/physionet-build/issues/1786 

> - I think this is an unrelated issue, but there is a really annoying bug that after editing an embargo (and perhaps other fields) drops the editor into a "project info" page that doesn't include editorial forms.

This appears to be by design. The form action is calling [`submission_info`](https://github.com/MIT-LCP/physionet-build/blob/3823cc97f072aeb51565e6fee34e4584df18aec6/physionet-django/console/views.py#L248) which doesn't include the editorial forms. Those forms are included under [`edit_submission`](https://github.com/MIT-LCP/physionet-build/blob/3823cc97f072aeb51565e6fee34e4584df18aec6/physionet-django/console/views.py#L306) in `views.py`.
